### PR TITLE
vendor electron-updater

### LINKED ISSUE #4213: Vendor desktop libraries
We do vendoring for web part for security reasons (libs directory). We should do the same for desktop-only libraries.

* [x] update-libs scripts needs to be updated to pull desktop deps
* [x] update debug and dist build to use the vendored files
* [x] create list of dependencies to pull
* [x] restrict dist build to only use `@tutao` dependencies from node_modules (node-resolve plugin)
* [ ] review the dependencies

## Test Notes
* [ ] auto-update works for mac, linux, windows
* [ ] offline mode works for mac, linux, windows
* [ ] saving credentials works for mac, linux, windows
* [ ] dev build of the desktop client works

### PR DESCRIPTION
**Pull request by ganthern**
it's now checked in and can be updated with buildSrc/updateLibs.js

close #4213

### CODE REVIEW

**Comment by ganthern** [comment]
still needs more work for copying the right version of keytar/better-sqlite3 at the right time and to the right place.

File: buildSrc/updateLibs.js [hunk 1]
```diff
@@ -59,3 +65,36 @@ async function roll(src, target) {
...
+/**
+ * specifically roll up electron-updater and all its dependencies into a single esm file.
+ * it's importing some electron internals directly, so we made a comprehensive list of
+ * exclusions to not roll up.
+ */
+async function rollUpdater(src, target) {
+	const bundle = await rollup({
+		input: path.join(__dirname, src),
+		external: [
+			"assert",
```
Comment (charlag) [inline buildSrc/updateLibs.js:80]: aren't most of these anyway node builtins that should be anyway not imported?

Comment (ganthern) [reply buildSrc/updateLibs.js:80]: rollup didn't think so apparently

File: buildSrc/buildWebapp.js [hunk 2]
```diff
@@ -32,7 +32,7 @@ export async function buildWebapp({ version, stage, host, measure, minify, proje
 	const polyfillBundle = await rollup({
 		input: ["src/polyfill.ts"],
 		plugins: [
-			typescript(),
+			typescript({ outputToFilesystem: true }),
```
Comment (charlag) [inline buildSrc/buildWebapp.js:35]: it is unclear why we do output to filesystem, maybe add a comment? (also I do not think it should be necessary?)

File: test/TestBuilder.js [hunk 3]
```diff
@@ -36,6 +36,23 @@ export async function runTestBuild({ clean, fast = false }) {
 	})
 	await runStep("Esbuild", async () => {
 		await esbuild({
+			// this is here because the test build targets esm and esbuild
+			// does not support dynamic requires, which better-sqlite3 uses
```
Comment (charlag) [inline test/TestBuilder.js:40]: should we just change it in our fork directly? maybe later tho

File: buildSrc/updateLibs.js [hunk 4]
```diff
@@ -42,6 +43,7 @@ async function run() {
 
 async function copyToLibs(files) {
 	for (let srcFile of files) {
+		console.log("transforming", srcFile.src || srcFile, "to", srcFile.target || "/libs")
```
Comment (charlag) [inline buildSrc/updateLibs.js:46]: log

File: buildSrc/DesktopBuilder.js [hunk 5]
```diff
@@ -125,32 +125,39 @@ async function rollupDesktop(dirname, outDir, version, platform, disableMinify)
...
-			keytarNativeBannerPlugin({
-				rootDir: projectRoot,
-				platform,
-			}),
 			nodeResolve({
 				preferBuiltins: true,
 				resolveOnly: [/^@tutao\/.*$/, "keytar"],
 			}),
-			// requireReturnsDefault: "preferred" is needed in order to correclty generate a wrapper for the native keytar module
+			// requireReturnsDefault: "preferred" is needed in order to correctly generate a wrapper for the native keytar module
```
Comment (charlag) [inline buildSrc/DesktopBuilder.js:142]: for keytar specifically it looks odd because we unpack it immediately anyway?

Comment (ganthern) [reply buildSrc/DesktopBuilder.js:142]: i just got that keytar is not cjs anymore now and that's probably why i had problems with that import

Comment (ganthern) [reply buildSrc/DesktopBuilder.js:142]: scratch that, `requireReturnsDefault` applies to cjs modules importing es modules, not the other way around

**Comment by ganthern** [comment]
We should probably think about forking keytar to @tutao and converting it to esm, it's barely doing anything on the JS side anyway.

**Review by charlag — APPROVED** [review]

**Pull request merged**